lib/src/resolve_context.cc

```diff
@@ -23,16 +23,47 @@ namespace hocon {
 
     resolve_result<shared_value> resolve_context::resolve(shared_value original, resolve_source const& source) const
     {
-        auto result = original->resolve_substitutions(*this, source);
-        auto resolved = result.value;
+        memo_key full_key {original, {}};
+        memo_key restricted_key = {nullptr, {}};
 
-        if (!resolved || resolved->get_resolve_status() == resolve_status::RESOLVED) {
-            return result;
-        }  else {
-            throw config_exception("Partial resolves and incomplete resolutions are not yet implemented");
+        shared_value cached;
+        auto cached_iter = _memos.find(full_key);
+        if (cached_iter != _memos.end()) {
+            cached = cached_iter->second;
         }
 
-        return result;
+        if (!cached && is_restricted_to_child()) {
+            restricted_key = {original, restrict_to_child()};
+            auto cached_iter = _memos.find(full_key);
+            if (cached_iter != _memos.end()) {
+                cached = cached_iter->second;
+            }
+        }
+        
+        if (cached) {
+            return make_resolve_result(*this, cached);
+        } else {
+            auto result = original->resolve_substitutions(*this, source);
+            auto resolved = result.value;
+            auto with_memo = result.context;
```

Comment:
Might want to make this a ref, so we're not copying the more often than we need to.